File: shared/src/main/java/org/securesystem/insular/analytics/AnalyticsImpl.java
package org.securesystem.insular.analytics;

import android.content.Context;
import android.os.Bundle;

import org.securesystem.insular.IslandApplication;
import org.securesystem.insular.shared.BuildConfig;

import org.intellij.lang.annotations.Pattern;

import javax.annotation.ParametersAreNonnullByDefault;

import androidx.annotation.CheckResult;
import androidx.annotation.Nullable;

/**
 * The analytics implementation in local process
 *
 * Created by Oasis on 2017/3/23.
 */
@ParametersAreNonnullByDefault
class AnalyticsImpl implements Analytics {

	@Override public @CheckResult Event event(final @Pattern("^[a-zA-Z][a-zA-Z0-9_]*$") String event) {
		final Bundle bundle = new Bundle();
		return new Event() {
			@Override public @CheckResult Event withRaw(final String key, final @Nullable String value) {
				if (value != null) bundle.putString(key, value);
				return this;
			}
			@Override public void send() { reportEvent(event, bundle); }
		};
	}

	@Override public Analytics trace(final String key, final String value) {
		return this;
	}

	@Override public Analytics trace(final String key, final int value) {
		return this;
	}

	@Override public Analytics trace(final String key, final boolean value) {
		return this;
	}

	@Override public void report(final Throwable t) {
	}

	@Override public void report(final String message, final Throwable t) {
	}

	@Override public void reportEvent(final String event, final Bundle params) {
	}

	@Override public void setProperty(final Property property, final String value) {
	}

	private AnalyticsImpl(final Context context) {

		// TODO: De-dup the user identity between Mainland and Island.
	}

	static Analytics $() {
		return sSingleton;
	}

	private static final AnalyticsImpl sSingleton = new AnalyticsImpl(IslandApplication.$());

	private static final String TAG = "Analytics";
}

File: shared/src/main/java/com/oasisfeng/island/analytics/AnalyticsImpl.kt
package com.oasisfeng.island.analytics

import android.content.Context
import android.os.Bundle
import androidx.annotation.CheckResult
import org.intellij.lang.annotations.Pattern

/**
 * The analytics implementation in local process
 *
 * Created by Oasis on 2017/3/23.
 */
internal class AnalyticsImpl(context: Context) : Analytics {

	@CheckResult override fun event(@Pattern("^[a-zA-Z][a-zA-Z0-9_]*$") event: String): Analytics.Event {
		val bundle = Bundle()
		return object : Analytics.Event {
			@CheckResult override fun withRaw(key: String, value: String?) = this.also { bundle.putString(key, value ?: return@also) }
			override fun send() = reportEvent(event, bundle) }
	}

	override fun trace(key: String, value: String): Analytics = this.also { CrashReport.setProperty(key, value) }
	override fun trace(key: String, value: Int): Analytics = this.also { CrashReport.setProperty(key, value) }
	override fun trace(key: String, value: Boolean): Analytics = this.also { CrashReport.setProperty(key, value) }

	override fun report(t: Throwable) = CrashReport.logException(t)
	override fun report(message: String, t: Throwable) { CrashReport.log(message); CrashReport.logException(t) }

	override fun reportEvent(event: String, params: Bundle) { }
	override fun setProperty(property: Analytics.Property, value: String) {	}
}

private const val TAG = "Analytics"
